### PR TITLE
`Programming exercises`: Allow users to configure the tab width in the online editor

### PR DESCRIPTION
**Pull request by b-fein**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
Changing the tab size is both important for students and tutors:
- Some programming languages do not use four spaces as default. Students should be able to easily use e.g., two spaces instead in the online editor.
- Some students might use proper TAB characters in their local editor. There it might be displayed in a width of eight. When the tutor then looks at the code in the online editor with a fixed tab size of four the formatting is pretty broken. By allowing the tutor to change how tabs are displayed even in read-only mode, the formatting as intended by the student can be displayed.

### Description
The code editor already has an option to set how tabs are displayed/how many spaces are inserted when pressing the TAB key in edit mode. This PR makes this setting available to the user.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Students
- 1 Programming Exercise

1. Submit to the programming exercise using the online editor. When changing the tab-width the editor should insert the corresponding number of spaces instead.
2. Checkout the repository locally and submit a file that contains actual TAB characters.
3. Open the submission as the instructor. Changing the tab size should change the display for the file with the actual TAB characters.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| `code-editor-ace.component.ts` | 56% | 70.3% |
| `ace-editor.component.ts` | 68% | 84.4% |

all new code covered 2x 100%

### Screenshots
Example in read-only mode when assessing submissions
Screenshot 2022-03-23 at 17-04-04 Programming Exercises
Screenshot 2022-03-23 at 17-04-16 Programming Exercises

Design of the settings menu changed since the screenshots above, effect of the change is still the same:
Screenshot 2022-04-05 at 17-41-50 Code Editor

### CODE REVIEW

File: src/main/webapp/i18n/de/editor.json [hunk 1]
```diff
@@ -42,6 +42,7 @@
             "loadingFiles": "Lade Dateien",
             "fileLoadedSuccess": "Datei erfolgreich geladen",
             "fetchingParticipation": "Deine Teilnahme wird geladen, bitte warte einen Moment...",
+            "tabSize": "Tabulatorweite",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/editor.json:45]: You're either missing a colon here or you accidentally inserted one in the englisch version

Comment (b-fein) [reply src/main/webapp/i18n/de/editor.json]: Good catch! Yes, there is supposed to be one here.

**Review by Dominik-Weinzierl — COMMENTED** [review]
The code changes look good to me and work as described. One suggestion: What do you think about adding minimal space? 

image

**Comment by b-fein** [comment]
@Dominik-Weinzierl Thanks for testing.

I changed it so that the filename and the tab configuration are only allowed to get this close to each other:
Screenshot 2022-04-01 at 10-44-46 Code Editor
Is this what you meant by adding minimal space?

**Review by Dominik-Weinzierl — DISMISSED** [review]
Perfect - thank you! Looks good to me :+1:

**Review by willergeorg — DISMISSED** [review]
Code lgtm. Tested with online editor and worked as expected. 👍

**Comment by krusche** [comment]
I was wondering if the UI needs to be so prominent (how often do you change the tab size?) or if we could add a settings icon in the upper right corner and display the edit UI in a popup?
Then the already complex UI would not get even more complex.

Would this be possible?

**Comment by b-fein** [comment]
@krusche That is possible. New design:
Screenshot 2022-04-05 at 17-41-50 Code Editor

As Eclipse uses TABs by default even for Java code for whatever reason, having to change the tab width during assignment of programming exercises actually happens quite often. I agree, though, that moving it to a drop-down is nicer.

**Comment by krusche** [comment]
Great work 👍🏻 thanks! exactly how I imagined it.

**Review by akesfeden — COMMENTED** [review]
Tested on TS1, tabs are displayed and modified as expected but I have a question: When I set the tab size with arrows in the input element max value (8) is enforced but when I type it I can set it arbitrarily high (see `//actual 2 tabs` in the screenshot). Do you think we should still validate typed input or should we allow high values?

image

**Comment by b-fein** [comment]
@akesfeden Good catch, I didn’t think about that. I now added a validation to replace the value with a valid one in range [1, 8] in case the user manually changed it to something else.

**Review by akesfeden — APPROVED** [review]
Thanks for the changes! Tested on TS1, works well

**Review by maximiliansoelch — APPROVED** [review]
Tested on TS3, works as described.
Code changes lgtm 👍

**Review by cremertim — APPROVED** [review]
great work

**Pull request merged**